File: core/src/main/java/org/elkd/core/client/TestClient.java
package org.elkd.core.client;

import com.google.common.util.concurrent.ListenableFuture;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.StatusRuntimeException;
import org.apache.log4j.Logger;
import org.elkd.core.server.ElkdClusterServiceGrpc;
import org.elkd.core.server.ElkdClusterServiceGrpc.ElkdClusterServiceFutureStub;
import org.elkd.core.server.RpcAppendEntriesRequest;
import org.elkd.core.server.RpcAppendEntriesResponse;
import org.elkd.core.server.RpcEntry;
import org.elkd.core.server.RpcStateMachineCommand;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import static org.elkd.core.server.RpcStateMachineCommand.Operation.SET;

public class TestClient {
  private static final Logger LOG = Logger.getLogger(TestClient.class);

  private final ManagedChannel mManagedChannel;
  private final ElkdClusterServiceFutureStub mStub;
  private final ExecutorService mExecutor = Executors.newSingleThreadExecutor();

  public TestClient(final String host, final int port) {
    this(ManagedChannelBuilder.forAddress(host, port).usePlaintext().build());
  }

  TestClient(final ManagedChannel managedChannel) {
    mManagedChannel = managedChannel;
    mStub = ElkdClusterServiceGrpc.newFutureStub(managedChannel);
  }

  public void shutdown() throws InterruptedException {
    mManagedChannel.awaitTermination(5, TimeUnit.SECONDS);
  }

  public void appendEntries() {
    LOG.info("appending entries call");

    final RpcAppendEntriesRequest request = RpcAppendEntriesRequest.newBuilder()
        .setTerm(0)
        .setPrevLogIndex(1)
        .setPrevLogTerm(1)
        .setLeaderId("node-1")
        .setLeaderCommit(0)
        .addEntries(
            RpcEntry.newBuilder().setEvent("amznStock").addCommands(
                RpcStateMachineCommand.newBuilder().setOperation(SET).setKey("price").setValue(String.valueOf(Math.random() * 2000)).build()
            ).build()
        )
        .build();
    ListenableFuture<RpcAppendEntriesResponse> response;
    try {
      response = mStub.appendEntries(request);
      response.addListener(() -> {
        try {
          LOG.info(response.get().getTerm());
        } catch (InterruptedException e) {
          e.printStackTrace();
        } catch (ExecutionException e) {
          e.printStackTrace();
        }
      }, mExecutor);
    } catch (final StatusRuntimeException e) {
      LOG.error(e);
    }
  }

  public static void main(String[] args) throws InterruptedException {
    final TestClient client = new TestClient("localhost", 9191);
    try {
      for (;;) {
        client.appendEntries();
        Thread.sleep((long) (Math.random() * 500));
      }
    } finally {
      client.shutdown();
    }
  }
}

File: core/src/main/java/org/elkd/core/runtime/client/TestClient.kt
package org.elkd.core.runtime.client

import io.grpc.ManagedChannelBuilder
import org.elkd.core.server.client.ElkdClientServiceGrpc
import org.elkd.core.server.client.RpcArgPair
import org.elkd.core.server.client.RpcClientRequest
import org.elkd.core.server.client.RpcClientResponse
import org.elkd.core.server.cluster.ElkdClusterServiceGrpc
import org.elkd.core.server.cluster.RpcLogTail
import org.elkd.core.server.cluster.RpcRequestVoteRequest
import java.util.concurrent.Executors
import java.util.concurrent.Future

fun main() {
  createTopic(9191)

//  requestVotesMultiple(9191)
}

private fun requestVotesMultiple(port: Int) {
  val stub = ElkdClusterServiceGrpc.newFutureStub(
      ManagedChannelBuilder.forAddress("localhost", port).usePlaintext().build()
  )

  val newFixedThreadPool = Executors.newFixedThreadPool(2)
  listOf("localhost:9292", "localhost:9393").map {
    RpcRequestVoteRequest.newBuilder()
        .setTerm(10)
        .setCandidateId(it)
        .addLogTails(RpcLogTail.newBuilder()
            .setTopicId("@syslog")
            .setLastLogIndex(1)
            .setLastLogTerm(1)
            .build())
        .build()
  }.map {
    stub.requestVote(it)
  }.forEach {
    val response = it.get()
    println("${response.term} ${response.voteGranted}")
  }

  Thread.sleep(5000)
}

private fun createTopic(port: Int) {
  val stub = ElkdClientServiceGrpc.newFutureStub(
      ManagedChannelBuilder.forAddress("localhost", port).usePlaintext().build()
  )

  var count = 0
  var future: Future<RpcClientResponse>
  do {
    future = stub.clientCommand(RpcClientRequest.newBuilder()
        .setCommand("create-topic")
        .addAllArgs(listOf(RpcArgPair.newBuilder()
            .setArg("namespace")
            .setParam("sensor_ambient")
            .build()))
        .build())
  } while (++count < 0)
  println(future.get().response)
}
